### PR TITLE
Condense `use` sections

### PR DESCRIPTION
**Pull request by nnethercote**
`use` items can be broken (or not) into sections in many ways: `crate` items, `self` items, `std` items, `rustc_*` items, third-party crate items. Also `pub` vs non-`pub`, and items with/without attributes (e.g. `#[cfg(parallel_compiler)]`. rustfmt is very unopinionated about this and, as a result, there is little consistency within the rustc codebase on the right way to do it.

*However*, for vanilla `use` items (i.e. no `pub` or attributes) it never makes sense for `use crate::a` to be in a different section to `use crate::b`, or `use rustc_foo::a` to be in a different section to `use rustc_bar::b`, etc. This PR fixes all such cases that I could find. I tried to each change in the simple/obvious way -- e.g. if there is a single `use std::foo` by itself in a section and then three `use std::...` items in another section then I moved the single one into the group of three -- but there were a few cases where there was no clear move and judgment was needed.

r? @RalfJung

### CODE REVIEW

**Comment by nnethercote** [comment]
@bors rollup=never p=1

Because this is highly conflict-prone.

**Comment by RalfJung** [comment]
The interpreter changes LGTM; I don't currently have the capacity to review the entire PR unfortunately.

r? compiler

**Comment by klensy** [comment]
Once i tried to use https://rust-lang.github.io/rustfmt/?version=master&search=#group_imports `StdExternalCrate` but that given truly big diff, so i given up :-(

**Comment by RalfJung** [comment]
All commits in Miri get mirrored over to rustc, so this isn't painting an adequate picture. Most of my work is on Miri.

I'm also trying to reduce the amount of time I spend on Rust, for my sanity's sake. It's not working out too well so far, but certainly I shouldn't increase the scope of what I review.

File: compiler/rustc_borrowck/src/lib.rs [hunk 1]
```diff
@@ -30,6 +30,13 @@ use rustc_middle::mir::*;
 use rustc_middle::query::Providers;
 use rustc_middle::ty::{self, ParamEnv, RegionVid, TyCtxt};
 use rustc_middle::{bug, span_bug};
+use rustc_mir_dataflow::impls::{
+    EverInitializedPlaces, MaybeInitializedPlaces, MaybeUninitializedPlaces,
+};
+use rustc_mir_dataflow::move_paths::{InitIndex, MoveOutIndex, MovePathIndex};
+use rustc_mir_dataflow::move_paths::{InitLocation, LookupResult, MoveData};
+use rustc_mir_dataflow::Analysis;
+use rustc_mir_dataflow::MoveDataParamEnv;
```
Comment (Zalathar) [inline compiler/rustc_borrowck/src/lib.rs:39]: Two pairs of same-module imports that can potentially be merged.

File: compiler/rustc_metadata/src/rmeta/mod.rs [hunk 2]
```diff
@@ -18,10 +15,13 @@ use rustc_hir::definitions::DefKey;
 use rustc_hir::lang_items::LangItem;
 use rustc_index::bit_set::BitSet;
 use rustc_index::IndexVec;
+use rustc_macros::{Decodable, Encodable, TyDecodable, TyEncodable};
 use rustc_macros::{MetadataDecodable, MetadataEncodable};
```
Comment (Zalathar) [inline compiler/rustc_metadata/src/rmeta/mod.rs:19]: Another pair of same-module imports here.

File: compiler/rustc_span/src/lib.rs [hunk 3]
```diff
@@ -38,6 +38,10 @@ extern crate self as rustc_span;
 #[macro_use]
 extern crate tracing;
 
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::stable_hasher::HashingControls;
+use rustc_data_structures::stable_hasher::{Hash128, Hash64, HashStable, StableHasher};
```
Comment (Zalathar) [inline compiler/rustc_span/src/lib.rs:43]: Another pair of same-module imports here.

**Comment by saethlin** [comment]
I am fine with moving entire lines around, but I do not want to see changes that turn this style:
```rust
use rustc_crate::a;
use rustc_crate::b;
```
Into this:
```rust
use rustc_crate::{a, b};
```
The latter style is extremely prone to merge conflicts, because any two PRs that want to add or remove an import from `rust_crate` will conflict with each other. Importing each item on its own line uses a considerable amount of space at the top of each file for imports, but really does result in less merge conflicts.

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 4]
```diff
@@ -13,31 +13,30 @@ use super::Selection;
...
 use rustc_middle::ty::fold::TypeFoldable;
 use rustc_middle::ty::visit::{MaxUniverse, TypeVisitable, TypeVisitableExt};
 use rustc_middle::ty::{self, Term, ToPredicate, Ty, TyCtxt};
+use rustc_middle::{bug, span_bug};
 use rustc_span::symbol::sym;
 
+use crate::errors::InherentProjectionNormalizationOverflow;
+use crate::infer::{BoundRegionConversionTime, InferOk};
+use crate::traits::normalize::normalize_with_depth;
+use crate::traits::normalize::normalize_with_depth_to;
```
Comment (Zalathar) [inline compiler/rustc_trait_selection/src/traits/project.rs:36]: Another pair of same-module imports here.

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 5]
```diff
@@ -42,6 +42,7 @@ use rustc_middle::dep_graph::DepNodeIndex;
 use rustc_middle::mir::interpret::ErrorHandled;
 use rustc_middle::ty::_match::MatchAgainstFreshVars;
 use rustc_middle::ty::abstract_const::NotConstEvaluatable;
+use rustc_middle::ty::print::with_no_trimmed_paths;
 use rustc_middle::ty::print::PrintTraitRefExt as _;
```
Comment (Zalathar) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:46]: Another pair of same-module imports here.

**Comment by nnethercote** [comment]
> I am fine with moving entire lines around, but I do not want to see changes that turn this style:
> 
> ```rust
> use rustc_crate::a;
> use rustc_crate::b;
> ```
> 
> Into this:
> 
> ```rust
> use rustc_crate::{a, b};
> ```

Huh, I didn't expect that. Just to make sure I'm understanding, in general your preferred style would be for something like this:
```rust
use rustc_span::{sym, Span, SpanDecoder, SpanEncoder, Symbol, DUMMY_SP};
```
To instead be this:
```rust
use rustc_span::sym;
use rustc_span::Span;
use rustc_span::SpanDecoder;
use rustc_span::SpanEncoder;
use rustc_span::Symbol;
use rustc_span::DUMMY_SP;
```
Is that right? (I understand you're not advocating for proactively splitting up existing lines, but over time as natural `use` churn occurs is that the style you would prefer to emerge?)

**Comment by saethlin** [comment]
> Is that right? (I understand you're not advocating for proactively splitting up existing lines, but over time as natural use churn occurs is that the style you would prefer to emerge?)

Yes you understand me completely.

**Comment by nnethercote** [comment]
Huh. The one-import-per-line style is *very* far from the current state of rustc!

So on one side I have @saethlin asking for no merging of mergeable lines, and on the other side I have @Zalathar pointing out all the places where I missed merging mergeable lines.

rustfmt really missed a trick in allowing so much flexibility here :(

**Comment by saethlin** [comment]
_as if on cue_

**Comment by RalfJung** [comment]
I did have to fix my share of merge conflicts in import lines... I don't think it is so bad as to warrant splitting all the imports, which makes them a lot more verbose and harder to parse.

We *could* do this
```rust
use rustc_span::{
    sym,
    Span,
    SpanDecoder,
    SpanEncoder,
    Symbol,
    DUMMY_SP,
};
```
to reduce verbosity and merge conflicts, but rustfmt won't let us.

**Comment by nnethercote** [comment]
> but rustfmt won't let us.

That is [imports_layout="Vertical"](https://rust-lang.github.io/rustfmt/?version=master&search=#imports_layout), AFAICT.

**Comment by nnethercote** [comment]
> as if on cue

You are right that one-import-line greatly reduces the likelihood of merge commits. But multiple-imports-per-line is the style used in the overwhelming majority of cases in the compiler's existing code. It feels odd to have changes rejected for conforming to the predominant style.

It would be **really nice** if imports_layout and group_imports were stable and reliable and we could just choose a value for each and let this be handled automatically.

**Comment by RalfJung** [comment]
> That is [imports_layout="Vertical"](https://rust-lang.github.io/rustfmt/?version=master&search=#imports_layout), AFAICT.

Ah, TIL! Thanks.

**Comment by saethlin** [comment]
> It feels odd to have changes rejected for conforming to the predominant style.

Has the team ever agreed on a style? I'm willing to +1 the majority opinion (as opposed to consensus), I'm just not aware of any discussion of the matter.

**Comment by nnethercote** [comment]
There is no style guide because `rustfmt` and `tidy` theoretically avoid the need for one, though theory and practice don't align entirely, as this PR indicates.

When I say "predominant" style I just mean what the existing code does. When it comes to section ordering, for example, there are a variety of styles and no single one dominates. But within sections, multiple imports per line is overwhelmingly common. Single import per line is extremely rare, and you can see in this PR both @Zalathar and I naturally gravitated to merging imports where possible. And the PR description is premised on multiple imports per line being non-controversial.

**Comment by saethlin** [comment]
I'm not going to sign off on changes which I think are net-negative on the basis of "it looks like that's how things are being done right now" as opposed to a conscious team decision. I've seen way too many things like this result in an outcome the team does not like because individuals keep going with "it looks like this is how we do things".

If you want someone to sign off on this, re-roll for another reviewer or start a team discussion.

**Comment by nnethercote** [comment]
> as if on cue

You are right that one-import-line greatly reduces the likelihood of merge commits. But multiple-imports-per-line is the style used in the overwhelming majority of cases in the compiler's existing code. It feels odd to have changes rejected for conforming to the predominant style.

It would be **really nice** if imports_layout and group_imports were stable and reliable and we could just choose a value for each and let this be handled automatically.

**Pull request closed without merging**